subsys/bluetooth/audio/tbs.c

```diff
@@ -2944,54 +2968,49 @@ int bt_tbs_set_uri_scheme_list(uint8_t bearer_index, const char **uri_list, uint
 		}
 
 		len += strlen(uri_list[i]);
+
 		if (len > sizeof(uri_scheme_list) - 1) {
 			return -ENOMEM;
+		} else if (strstr(inst->uri_scheme_list, uri_list[i]) != NULL) {
+			len -= strlen(uri_list[i]);
+			LOG_DBG("uri scheme %s is already added to the list", uri_list[i]);
+			continue;
+		} else {
+			/* Store list in temp list in case something goes wrong */
+			strcat(uri_scheme_list, uri_list[i]);
 		}
-
-		/* Store list in temp list in case something goes wrong */
-		strcat(uri_scheme_list, uri_list[i]);
 	}
 
-	if (strcmp(inst->uri_scheme_list, uri_scheme_list) == 0) {
-		/* identical; don't update or notify */
+	if (len == 0) {
+		/* no new uri-scheme added; don't update or notify */
 		return 0;
 	}
 
+	size_t uri_len = strlen(inst->uri_scheme_list);
+
 	err = k_mutex_lock(&inst->mutex, K_NO_WAIT);
 	if (err != 0) {
 		LOG_DBG("Failed to lock mutex");
 		return -EBUSY;
 	}
 
+	if ((len + uri_len + 1) >= READ_BUF_SIZE) {
+		LOG_WRN("Cannot fit all URI scheme list");
+		return -ENOMEM;
+	}
```

Comment:
We already have a length check against `uri_scheme_list` (`CONFIG_BT_TBS_MAX_SCHEME_LIST_LENGTH`). 

`CONFIG_BT_TBS_MAX_SCHEME_LIST_LENGTH` is maximum 512, so we should be able to modify the `READ_BUF_SIZE` so that it takes `CONFIG_BT_TBS_MAX_SCHEME_LIST_LENGTH` into account. Perhaps it needs to add something like `(CONFIG_BT_TBS_BEARER_COUNT * CONFIG_BT_TBS_MAX_SCHEME_LIST_LENGTH + CONFIG_BT_TBS_MAX_SCHEME_LIST_LENGTH)` to also "fit" the case of GTBS, where we union the values.

Reply:
See my comment from https://github.com/zephyrproject-rtos/zephyr/pull/91833#discussion_r2154560552 - The function shall not _add_ but only _set_ the values.